clients/src/main/java/org/apache/kafka/clients/consumer/PartitionAssignor.java

```diff
@@ -0,0 +1,148 @@
+ * assignment decisions. For this, you can override {@link #subscription(Set)} and provide custom
+ * userData in the returned Subscription. For example, to have a rack-aware assignor, an implementation
+ * can use this user data to forward the rackId belonging to each member.
+ */
+public interface PartitionAssignor {
+
+    /**
+     * Return serialized data that will included in the serializable subscription object and can
+     * be leveraged in {@link #assign(Cluster, Map)} ((e.g. local host/rack information)
+     *
+     * @return Non-null subscription with optional user data
+     */
+    default ByteBuffer userData() {
+        return ByteBuffer.wrap(new byte[0]);
+    }
+
+    /**
+     * Perform the group assignment given the member subscriptions and current cluster metadata.
+     * @param metadata Current topic/broker metadata known by consumer
+     * @param subscriptions Subscriptions from all members provided through {@link #subscription(Set)}
+     * @return A map from the members to their respective assignment. This should have one entry
+     *         for all members who in the input subscription map.
+     */
+    GroupAssignment assign(Cluster metadata, GroupSubscription subscriptions);
+
+    /**
+     * Callback which is invoked when a group member receives its assignment from the leader.
+     * @param assignment The local member's assignment as provided by the leader in {@link #assign(Cluster, Map)}
+     */
+    void onAssignment(Assignment assignment);
+
+    /**
+     * Indicate which rebalance protocol this assignor works with;
+     * By default it should always work with {@link RebalanceProtocol#EAGER}.
+     */
+    default List<RebalanceProtocol> supportedProtocols() {
+        return Collections.singletonList(RebalanceProtocol.EAGER);
+    }
+
+    /**
+     * Return the version of the assignor which indicates how the user metadata encodings
+     * and the assignment algorithm gets evolved.
+     */
+    default short version() {
+        return (short) 0;
+    }
+
+    /**
+     * Unique name for this assignor (e.g. "range" or "roundrobin" or "sticky")
+     * @return non-null unique name
+     */
+    String name();
+
+    enum RebalanceProtocol {
+        EAGER((byte) 0), COOPERATIVE((byte) 1);
+
+        private final byte id;
+
+        RebalanceProtocol(byte id) {
+            this.id = id;
+        }
+
+        public byte id() {
+            return id;
+        }
+
+        public static RebalanceProtocol forId(byte id) {
+            switch (id) {
+                case 0:
+                    return EAGER;
+                case 1:
+                    return COOPERATIVE;
+                default:
+                    throw new IllegalArgumentException("Unknown rebalance protocol id: " + id);
+            }
+        }
+    }
+
+    final class GroupAssignment {
+        private final Map<String, Assignment> assignments;
```

Comment:
The annoying thing about this definition is that `Assignment` contains the version, but a user shouldn't have to know this.

Reply:
Seems like the ConsumerProtocol logic can encapsulate it so it doesn't even need to be a field at all?